src/Network/GRPC/MQTT/Proto.hs

```diff
@@ -0,0 +1,672 @@
+   in throwIO issue
+
+throwProtoIO :: ProtoIOErrorTag -> Turtle.FilePath -> IO a
+throwProtoIO tag filepath = throwIO (ProtoIOError filepath tag)
+
+-- Proto IO Errors - Error Tags ------------------------------------------------
+
+-- | An enumeration of error codes that 'ProtoIOError' is tagged with.
+--
+-- @since 0.1.0.0
+data ProtoIOErrorTag
+  = ProtoInvalidExtension
+  | ProtoFileDoesNotExist
+  | ProtoPathIsDirectory
+  deriving (Enum, Eq, Ord)
+
+-- | @since 0.1.0.0
+instance Show ProtoIOErrorTag where
+  show ProtoInvalidExtension = "filepath must have a *.proto extension"
+  show ProtoFileDoesNotExist = "proto filepath not found (does not exist)"
+  show ProtoPathIsDirectory = "filepath refers to a directory (expected a file)"
+  {-# INLINE show #-}
+
+-- Proto Data Casts ------------------------------------------------------------
+
+-- | Obtain the kind of 'DatumRep' used to represent a 'DotProtoValue'.
+--
+-- @since 0.1.0.0
+datumRepOf :: DotProtoValue -> DatumRep
+datumRepOf DotProto.BoolLit{} = DRepBoolean
+datumRepOf DotProto.FloatLit{} = DRepDecimal
+datumRepOf DotProto.Identifier{} = DRepIdent
+datumRepOf DotProto.IntLit{} = DRepInteger
+datumRepOf DotProto.StringLit{} = DRepString
+
+-- | Type-safe cast operation from a 'DotProtoValue' to the equivalent Haskell
+-- representation.
+--
+-- >>> castDatumRepWith @Bool DRepBoolean (DotProto.BoolLit False)
+-- Just False
+--
+-- The cast will fail returning 'Nothing' if the given 'DatumRep' does not match
+-- the representation of the 'DotProtoValue' and the expected result type @a@.
+--
+-- >>> -- Fails because a 'DRepBoolean' scalar cannot be cast to a 'String'.
+-- >>> castDatumRepWith @String DRepString (DotProto.BoolLit True)
+-- Nothing
+--
+-- @since 0.1.0.0
+castDatumRepWith :: forall a. Typeable a => DatumRep -> DotProtoValue -> Maybe a
+castDatumRepWith datrep val =
+  case datrep of
+    DRepBoolean -> do
+      Refl <- eqT @Bool @a
+      castValue val
+    DRepDecimal -> do
+      Refl <- eqT @Double @a
+      castValue val
+    DRepInteger -> do
+      Refl <- eqT @Int @a
+      castValue val
+    DRepString -> do
+      Refl <- eqT @String @a
+      castValue val
+    DRepIdent -> do
+      Refl <- eqT @DotProtoIdentifier @a
+      castValue val
+  where
+    castValue :: DotProtoValue -> Maybe a
+    castValue x = do
+      casted <- sequence (gmapQ cast x)
+      viaNonEmpty head casted
+
+-- Protobuf Data - Representation ----------------------------------------------
+
+-- | 'DatumRep' enumerates the /kinds/ of scalars that can be represented by a
+-- 'DotProtoValue' value.
+--
+-- @since 0.1.0.0
+data DatumRep
+  = -- | Boolean representations correspond to the @bool@ .proto type.
+    DRepBoolean
+  | -- | Decimal representations capture the @double@ and @float@ .proto types.
+    DRepDecimal
+  | -- | Integer representations capture all signed, unsigned, and fixed-width
```

Comment:
Why is conflating these many different integer representations into one, with the `castDatumRepWith` function, safe?

I have the same question for the `DRepDecimal` and `DRepString` tags.

Reply:
However, in conflating `DRepDecimal` and `DRepInteger` representations, there is nothing preventing a situation where `castDatum` can overflow or underflow in the process of casting `Int` to `Int8`. In these cases though, the casts will inherit the overflow/underflow behavior of `fromIntegral`.

Reply:
The point you make in your second comment is what I meant by asking why it was safe, you're relying on `fromIntegral`'s behavior which could produce an overflow. I don't think this is safe and instead you should explicate the different integer and decimal types in DRep so that you have a 1:1 conversion to their counterparts in Haskell (and the conversions can happen safely).

Reply:
I guess I'm just looking at it from this side of the `proto3-suite` API boundary. Here you will only ever see an `Int` inside a `DotProtoValue`.

I agree `Integer` would be a safe type to use for `IntLit`, but that seems like something that would need to be addressed within `proto3-suite`.

Reply:
I understand now @Friede80 and @riz0id and agree, without a `ProtoDatum Int8` instance, this is fine. I'm still worried about converting from a Double to a Float because that is lossy.